### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by cef62**
This PR is related to #10 

The PR add gitbook and all its dependencies and configurations.
_important_ to view the output on the local server the first time the docs must compiled before `npm run docs:build` to correctly install gitbook plugins.

@MicheleBertoli I've set up placeholders for main sections we discussed. Feel free to work and push directly on this branch while we work on the documentation.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+For a non-visual library this is a trivial problem to solve. For a component library using `css-modules` and `postcss` plugins is a real problematic situation. The fast and dirty solution is to consume directly the library source code from the consumer project, this approach bring several drawbacks: the consumer project must knows specific details on how to compile the library and must have all the library de-dependencies installed. It can be done but in th end is really better avoid separation between library and final project.
```
Comment (MicheleBertoli) [inline docs/README.md]: > in th end

[...] better avoid separation [...]
I guess you wanted to say the opposite, right?

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+The ideal solution is to compile the `css-modules` and consume the `es2015` modules without css imports. Currently there are really few toolsthat compile `css-module` extracting and decoupling them from the javascript modules and none of this tool allows the flexibility we want.
```
Comment (MicheleBertoli) [inline docs/README.md]: > toolsthat

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+## What css-module-compiler can do for you
+
+The compiler is a small autonomous node module tht can be used programmatically from javascript or directly from the CLI. To use the module is enough invoke it passing a source folder, the folder will be traversed and all css files will be compiled as `css-modules`, all the generated css files will be merged, deduped and optimized in a single css files. All `es2015` modules will be checked using an AST parser and all `css` import declaration will be substituted with a static object containing the generated css classnames.
```
Comment (MicheleBertoli) [inline docs/README.md]: > tht

**Comment by MicheleBertoli** [comment]
Well done!

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# CSS Modules Compiler
+
+> Compile css-modules to static css and remove non standard code from javascript files
+
+This module started to support a real project using [css-modules](https://github.com/css-modules/css-modules), [postcss](https://github.com/postcss/postcss) and [webpack](https://webpack.github.io/) to produce a library of reusable components. The project goal is to expose a rich set of components offering an easy way for other projects to use the desired components benefiting from the [tree-shaking](https://medium.com/@roman01la/dead-code-elimination-and-tree-shaking-in-javascript-build-systems-fb8512c86edf#.krdhto4gs) offered by modern bundlers. To achieve the goal the library must expose directly `es2015` modules, without usage of transpilers and file concatenation.
```
Comment (MicheleBertoli) [inline docs/README.md]: started => has been created to

File: docs/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+For a non-visual library this is a trivial problem to solve. For a component library using `css-modules` and `postcss` plugins is a real problematic situation. The fast and dirty solution is to consume directly the library source code from the consumer project, this approach bring several drawbacks: the consumer project must knows specific details on how to compile the library and must have all the library de-dependencies installed. It can be done but in the end is really better search for a clear separation between library and final project.
```
Comment (MicheleBertoli) [inline docs/README.md:7]: knows => know

search => to aim

File: docs/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+The ideal solution is to compile the `css-modules` and consume the `es2015` modules without css imports. Currently there are really few tools that compile `css-module` extracting and decoupling them from the javascript modules and none of this tool allows the flexibility we want.
+
+## What css-module-compiler can do for you
+
+The compiler is a small autonomous node module that can be used programmatically from javascript or directly from the CLI. To use the module is enough invoke it passing a source folder, the folder will be traversed and all css files will be compiled as `css-modules`, all the generated css files will be merged, deduped and optimized in a single css files. All `es2015` modules will be checked using an AST parser and all `css` import declaration will be substituted with a static object containing the generated css classnames.
```
Comment (MicheleBertoli) [inline docs/README.md]: invoke => to invoke

all css files => all the css files

in a single css files => in a single css file

all es2015 => all the es2015

all css import => all the css import

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+The compile accept several options: postcss plugins to be used when compiling css files, a blacklist of pattern used to avoid compilation of non `css-modules` files, a target folder to duplicate the source and avoid changing the original sources and more.
```
Comment (MicheleBertoli) [inline docs/README.md]: The compile accept => The compile command accepts

pattern => patterns

File: docs/api.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# API
+
+Currently the module exports a single public method:
+
+## compileCss(source:string, options:object)
+
+The method accept a source folder and a set of options to compile all css-modules found into a single css file, removing css imports declaration and substituting them with static objects of css class names.
```
Comment (MicheleBertoli) [inline docs/api.md]: accept => accepts

File: docs/api.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+### - source : string
+
+Relative or absolute path to the source folder containing the css/js to be processed.
+
+### - options : object
+
+* **`targetFolder`** *string* Relative or absolute path to the target folder. The source folder will be copied to the given path and its contents will be compiled.
```
Comment (MicheleBertoli) [inline docs/api.md]: contents => content

File: docs/api.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+* **`targetName`** *string* Name used to create the final compiled `css` output, the extension is required. Eg. `final-output.css`
+* **`plugins`** *array* List of **postcss plugins** instances ready to be passed to postcss. Eg. `[atImport(), colorFunction()]`
+* **`blacklist`** *array* List of **glob pattern** used to filter some css files from css compilation. Eg. `['*.global.css', 'main.css']`
```
Comment (MicheleBertoli) [inline docs/api.md]: filter => filter out

File: docs/cli.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+Usage
+    css-modules-compiler.js -s <source folder> -t <target folder> -n <output name> -p <plugin modules>
+
+Options:
+  --help                     Show help [boolean]
+  -s, --source               Source folder to be processed, can be a relative or an absolute path. [string] [required]
+  -t, --target               Target folder, if defined the source folder will be copied to the given path and the modules will process the duplicated folder. [string]
+  -p, --plugins              Space separated string of names of npm postcss
+                             plugins. This parameter must be defined as last. [array]
+  --plugins-config, --pconf  Path to a node module invoked at runtime to receive a list of plugins.The modules must export a function. [string]
```
Comment (MicheleBertoli) [inline docs/cli.md]: .The => . The

File: docs/cli.md [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  -n, --name                 Name for the generated css file, default to styles.css [string] [default: "styles.css"]
```
Comment (MicheleBertoli) [inline docs/cli.md]: default => defaults

File: docs/setup.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+```
+
+Once is installed you're ready to go.
+
+## Usage
+
+The module can be
+
+* invoked from `cli`
+* invoked from an npm script
```
Comment (MicheleBertoli) [inline docs/setup.md]: an npm => a npm

File: docs/README.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+# CSS Modules Compiler
+
+> Compile css-modules to static css and remove non standard code from javascript files
```
Comment (MicheleBertoli) [inline docs/README.md]: JavaScript

File: docs/README.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+For a non-visual library this is a trivial problem to solve. For a component library using `css-modules` and `postcss` plugins is a real problematic situation. The fast and dirty solution is to consume directly the library source code from the consumer project, this approach bring several drawbacks: the consumer project must knows specific details on how to compile the library and must have all the library de-dependencies installed. It can be done but in the end is really better search for a clear separation between library and final project.
+The ideal solution is to compile the `css-modules` and consume the `es2015` modules without css imports. Currently there are really few tools that compile `css-module` extracting and decoupling them from the javascript modules and none of this tool allows the flexibility we want.
```
Comment (MicheleBertoli) [inline docs/README.md]: JavaScript

File: docs/README.md (hunk 6 above)
Comment (MicheleBertoli) [inline docs/README.md]: JavaScript

File: docs/setup.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+* executed directly from nodeJS modules or gulp tasks
```
Comment (MicheleBertoli) [inline docs/setup.md]: Node.js

**Comment by cef62** [comment]
@MicheleBertoli I've fixed the docs following your suggestions and I've completed `faq` and `known issues`.

When merged with your `quickstart` branch docs should be ready to be published. I'll wait to add you complete your branch to evaluate if we need more examples.

File: docs/known-issues.md [hunk 17]
```diff
@@ -1 +1,8 @@
-TBD
+# Known Issues
+
+The compiler currently have some known limitation that we are working to resolve.
+
+* [ ] The compiler doesn't accept more than one source folder
+* [ ] Compiling *css* files using `composes` to import declarations from another *css* file that uses **postcss** `imports` doesn't work
```
Comment (MicheleBertoli) [inline docs/known-issues.md]: what about?

> composed CSS files are not processed by **postcss** which means that, for example, `@import` statements don't get replaced

File: docs/faq.md [hunk 18]
```diff
@@ -1 +1,13 @@
...
+
+Currently is not possible to use the module as plugin for bundlers tools. We aim to support webpack as soon as possible, support for other tools will follow.
+
+### Can I use the module with Gulp
+
+Yes but partially. `ccs.modules-compiler` is a simple node modules and thus can be used inside any gulp task. The module at the moment is not available as Gulp plugin therefore can't be used inside a Gulp stream.
+
+### Which version of Node.js is supported
+
+Currently we support only `Node 6.x` and higher because the module is written using several `es2015` features.
```
Comment (MicheleBertoli) [inline docs/faq.md]: and higher => or higher

File: docs/faq.md [hunk 19]
```diff
@@ -1 +1,13 @@
-TBD
+# FAQ
+
+### Can I use the module with other bundler (Webpack, Rollup, browserify, etc...)
+
+Currently is not possible to use the module as plugin for bundlers tools. We aim to support webpack as soon as possible, support for other tools will follow.
+
+### Can I use the module with Gulp
+
+Yes but partially. `ccs.modules-compiler` is a simple node modules and thus can be used inside any gulp task. The module at the moment is not available as Gulp plugin therefore can't be used inside a Gulp stream.
```
Comment (MicheleBertoli) [inline docs/faq.md]: modules => module
gulp => Gulp

File: docs/faq.md [hunk 20]
```diff
@@ -1 +1,13 @@
-TBD
+# FAQ
+
+### Can I use the module with other bundler (Webpack, Rollup, browserify, etc...)
```
Comment (MicheleBertoli) [inline docs/faq.md]: bundler => bundlers
browserify => Browserify

File: docs/faq.md [hunk 21]
```diff
@@ -1 +1,13 @@
...
+
+Currently is not possible to use the module as plugin for bundlers tools. We aim to support webpack as soon as possible, support for other tools will follow.
```
Comment (MicheleBertoli) [inline docs/faq.md]: Webpack

**Comment by MicheleBertoli** [comment]
Wow @cef62, you are awesome!
As I wrote in #27 I think the Quick Start and the Examples are fine for now.
We should merge, let some people play with it, and improve the docs according to the feedback we will receive.
What do you think?

File: docs/known-issues.md (hunk 17 above)
Comment (cef62) [inline docs/known-issues.md]: Way better than mine 😄

**Comment by cef62** [comment]
Thanks again for the english check @MicheleBertoli 👏 

I totally agree with you, I'll merge both PRs right now! 😎

**Pull request merged**